### ISSUE TITLE
tape.batch_jacobian gives different results if changing parallel_iterations

### ISSUE THREAD

**Issue by PythonFZ** [tensorflow/tensorflow#47785]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 x64
- TensorFlow installed from (source or binary): binary (pip install)
- TensorFlow version (use command below): 2.4.1
- Python version: 3.8.5
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: -
- GPU model and memory: -

**Describe the current behavior**

Using [batch_jacobian](https://www.tensorflow.org/api_docs/python/tf/GradientTape#batch_jacobian) with different values for `parallel_iterations` leads to differences in the output and the computational speed above the expected numerical errors.

**Describe the expected behavior**

Using `tape.batch_jacobian` should result in mostly the same outputs independent of `parallel_iterations`.

In my custom Code (which I did not provide) I encountered differences in calculation speed from 500 us for `parallel_iterations = 3` up to 238 ms for `parallel_iterations=8` without any pattern that could be related to the `parallel_iterations`.

I was able to reproduce this behaviour with a simplified version using `tf.einsum` as a function which provides a jacobian.
The error becomes zero when `parallel_iterations >=16`.

**Standalone code to reproduce the issue**

```py
import tensorflow as tf
data = tf.random.normal((32, 16, 16))

def func(r):
    return tf.einsum("bij, bij -> bi", r, r)

@tf.function
def func_grad(r, p_it = 2):
    with tf.GradientTape() as tape:
        tape.watch(r)
        out = func(r)
    return tape.batch_jacobian(out, r, unconnected_gradients="zero", parallel_iterations=p_it)

def func_grad_no_parallel(r, p_it = 2):
    with tf.GradientTape(persistent=True) as tape:
        tape.watch(r)
        out = func(r)
    return tape.batch_jacobian(out, r, experimental_use_pfor=False)

def error_analysis(y_true, y_second):
    rmse = tf.sqrt(tf.reduce_mean(tf.square(y_true - y_second)))
    max_err = tf.reduce_max(y_true - y_second)
    print(f"RMSE: {rmse} \t MAX: {max_err}")

grad_out = func_grad_no_parallel(data)
for x in range(2, 24):
    print(f"Parallel iterations: {x}")
    jacobian_out = func_grad(data, p_it=x)
    error_analysis(grad_out, jacobian_out)
    %timeit func_grad(data, p_it = x)
    print('-----------------------------------')
```
```
Parallel iterations: 8
RMSE: 0.5038933157920837 	 MAX: 7.24960470199585
1000 loops, best of 5: 667 µs per loop
-----------------------------------
Parallel iterations: 9
RMSE: 0.3786795139312744 	 MAX: 7.24960470199585
1000 loops, best of 5: 1.13 ms per loop
-----------------------------------
Parallel iterations: 16
RMSE: 0.5038933157920837 	 MAX: 7.24960470199585
1000 loops, best of 5: 645 µs per loop
-----------------------------------
Parallel iterations: 17
RMSE: 0.0 	 MAX: 0.0
1000 loops, best of 5: 1.03 ms per loop
-----------------------------------
```

**Other info / logs**

I was comparing the computation of the jacobian with a persisent GradientTape and looping over the last dimension of my data to get the individual gradients. Without using the `parallel_iterations` the loop over gradients and the jacobian are identical, but computing the jacobian is up to 4x times slower. When using `parallel_iterations` it can get more than twice as fast but comes with large errors/differences to the gradient loop method.

I added some simplified version of this approach to the end of the gist as well. This is not directly related to the bug but I thought, that using the Jacobian would be more efficient than unstacking and looping over the Tensor.

```py

import tensorflow as tf
data = tf.random.normal((32, 16, 3))

def func(r):
    r_mat = r[:, :, tf.newaxis, :] - r[:, tf.newaxis, :, :]
    return tf.einsum("bijk, bijk -> bi", r_mat, r_mat)

@tf.function
def func_grad_no_parallel(r):
    with tf.GradientTape(persistent=True) as tape:
        tape.watch(r)
        out = func(r)
    return tape.batch_jacobian(out, r, experimental_use_pfor=False)

@tf.function
def test_grad(r):
    with tf.GradientTape(persistent=True) as tape:
        tape.watch(r)
        out = func(r)
        out = tf.unstack(out, axis=1)
    deriv = []
    for frame in out:
        deriv.append(tape.gradient(frame, r))
    return tf.stack(deriv, axis=1)

tf.reduce_sum(func_grad_no_parallel(data) - test_grad(data))  # they are equal (without parallel_iterations)

%timeit func_grad_no_parallel(data)  # for this simplified version is still slower
%timeit test_grad(data)

```

You can find a [GIST here](https://colab.research.google.com/drive/1h42fHAeWvDn5gRlU5BBWqwHmuBkHQHg5?usp=sharing)

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3, TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/954bca9237e735e977a8868367be27af/47785.ipynb#scrollTo=Qd1dl5KrcQRb). Thanks!

**Comment by allenlavoie** [comment]
Thank you for the report. Looks like a capturing issue, where the gradient of the placeholder in the while_loop isn't recorded properly as being associated with the external tensor being captured. So `tape.gradient` returns None and the wrapper code crashes, or with `unconnected_gradients="zero"` it returns an incorrect result. The cases where the result is correct correspond to vectorized_map not using a while_loop.

I'll see if I can reproduce it with just a regular while_loop capturing a watched tensor.

**Comment by allenlavoie** [comment]
Looks like this is actually because the tape needs `persistent=True`, since vectorized_map calls the function that computes gradients twice. I'll add an explicit error message, but in the meantime I'd use `GradientTape(persistent=True)` in your code for `batch_jacobian(..., parallel_iterations=...)`. Sorry you ran into this.

**Comment by PythonFZ** [comment]
Thanks a lot! Adding `persistent=True` solved all the issues I had.
